rfcs/GraphQLOverWebSocket.md

```diff
@@ -0,0 +1,191 @@
+
+**All** messages contain the `type` field outlining the action this message describes. Depending on the type, the message can contain two more _optional_ fields:
+
+- `id` used for uniquely identifying server responses and connecting them with the client's requests
+- `payload` holding the extra "payload" information to go with the specific message type
+
+The server can close the socket (kick the client off) at any time. The close event dispatched by the server is used to describe the fatal error to the client.
+
+The client closes the socket and the connection by dispatching a `1000: Normal Closure` close event to the server indicating a normal closure.
+
+## Message types
+
+### `ConnectionInit`
+
+Direction: **Client -> Server**
+
+Indicates that the client wants to establish a connection within the existing socket. This connection is **not** the actual WebSocket communication channel, but is rather a frame within it asking the server to allow future operation requests.
+
+The server must receive the connection initialisation message within the allowed waiting time specified in the `connectionInitWaitTimeout` parameter during the server setup. If the client does not request a connection within the allowed timeout, the server will close the socket with the event: `4408: Connection initialisation timeout`.
+
+If the server receives more than one `ConnectionInit` message at any given time, the server will close the socket with the event `4429: Too many initialisation requests`.
+
+```typescript
+interface ConnectionInitMessage {
+  type: 'connection_init';
+  payload?: Record<string, unknown>;
+}
+```
+
+### `ConnectionAck`
+
+Direction: **Server -> Client**
+
+Expected response to the `ConnectionInit` message from the client acknowledging a successful connection with the server.
+
+```typescript
+interface ConnectionAckMessage {
+  type: 'connection_ack';
+}
+```
+
+The client is now **ready** to request subscription operations.
+
+### `Subscribe`
+
+Direction: **Client -> Server**
+
+Requests an operation specified in the message `payload`. This message provides a unique ID field to connect published messages to the operation requested by this message.
+
+If there is already an active subscriber for a streaming operation matching the provided ID, the server will close the socket immediately with the event `4409: Subscriber for <unique-operation-id> already exists`. The server may not assert this rule for operations returning a single result as they do not require reservations for additional future events.
+
+```typescript
+import { DocumentNode } from 'graphql';
+
+interface SubscribeMessage {
+  id: '<unique-operation-id>';
+  type: 'subscribe';
+  payload: {
+    operationName?: string | null;
+    query: string | DocumentNode;
```

Comment:
I have raised a [question](https://github.com/enisdenjo/graphql-ws/issues/44) if the protocol should support `DocumentNode` as a type of query and as @enisdenjo  suggested, it would be nice to collect more opinions on this.

So, what's going on... I think that for other languages than JS it is hard to use the `DocumentNode` in the server at the moment as its JS specific representation of GQL query AST. Yes, all GraphQL implementations use some AST representation, but its rather internal implementation detail and it isn't usually part of library public API (at least in [GraphQL java](https://github.com/graphql-java/graphql-java)). Moreover, I think there already exists a platform-independent compact query representation as a string, but there are probably some benefits from sending serialized AST as JSON. As I understood, the main concerns are about performance. I also think that an alternative format (e.g. `DocumentNode` as JSON) would be well-defined in this specification to be easy to implement in any language.

For more details see:
- https://github.com/enisdenjo/graphql-ws/issues/44
- https://github.com/enisdenjo/graphql-ws/pull/45

I'm looking forward to your opinions.

Reply:
It is indeed true that `DocumentNode` is very language-specific implementation, so it should not be in this spec. Many of the concerns are already raised in https://github.com/enisdenjo/graphql-ws/pull/45. 

Performance-wise, it can do more harm then help. Instead, the server implementation can choose to cache the DocumentNode using `query string`-`DocumentNode` pair cache after parsing and validation

Reply:
Great insight from @hoangvvo, its much better to introduce caching in comparison to doing some weird optimisations by skipping the parse step.

Nevertheless, I agree with all statements above and am willing to change the Protocol to reflect that.

However, I'd like to hear what others this about this before I make the change. @abernix @balshor @benjie @deinok @erikwittern @jaydenseric @michaelstaib @mike-marcacci @mmatsa @sjparsons @spawnia @sungam3r feedback would be much appreciated.

To reduce movement, here is the description from https://github.com/enisdenjo/graphql-ws/pull/45:
> Because of parity with HTTP conventions and reasons such as:
> - Size _(the serialised `Document` object can be much bigger than a simple query)_
> - Security _(no objects allowed through the `query` field, must provide parsable query)_
> - Implementation independent _(#44)_
> 
> it might make sense to strictly require the `query` field to be of a string type.
> 
> **However**, there are reasons where we want to support the `Document` object too:
> - Performance _(you might have a ready `Document` to be sent from the client to avoid query parsing altogether)_
> - You might want to use [`graphql-tag`](https://github.com/apollographql/graphql-tag) and get tooling support _(like from Prettier or syntax highlighting)_
> - You use the [Apollo Client GraphQLRequest type](https://github.com/apollographql/apollo-client/blob/b878385257223d1e45479033ff5dc5e3a4f7a4ab/src/link/core/types.ts#L6-L12). Like in [the WS link](https://github.com/apollographql/apollo-client/blob/e145a860a10c12fdc4d863379f32e4080597323b/src/link/ws/index.ts#L51)
> - You use the [Urql GraphQLRequestType](https://github.com/FormidableLabs/urql/blob/db7412ea56358dc057b225e6f13713ca16e27173/packages/core/src/types.ts#L25-L31)
> - The [RequestDocument from `graphql-request` supports both types too](https://github.com/prisma-labs/graphql-request/blob/c75a29a9a2a177b0cddb41718b333fb14c9d9917/src/types.ts#L56)
> - The [GraphQL over HTTP Proposal spec](https://github.com/graphql/graphql-over-http/blob/master/spec/GraphQLOverHTTP.md#request-parameters) describes the `query` request parameter as:
>   > A Document containing GraphQL Operations and Fragments to execute.

Reply:
When transmitted over the wire, the `query` (operation) should be the string representation in GraphQL language. Other things should be possible (e.g. for persisted operations/persisted queries, you might instead send a `documentId`) but the spec should relate to the standard cross-platform representation which is the GraphQL string. What the library itself accepts, however, should be unconstrainted by this. Passing the DocumentNode to the library is perfectly valid, so long as the library `graphql.print(ast)`'s it and sends the string representation over the wire.

Reply:
Perfect! I will merge the PR then. Because my lib is intended to be another tool in your workspace and in order to keep it secure, small and lean - the library will drop the `DocumentNode` support from the message too.